p2p/http/libp2phttp.go

```diff
@@ -0,0 +1,756 @@
+	// connections for this server, when in fact they are closing all idle
+	// connections
+}
+
+type namespacedRoundTripper struct {
+	http.RoundTripper
+	protocolPrefix    string
+	protocolPrefixRaw string
+}
+
+func (rt *namespacedRoundTripper) GetPeerMetadata() (PeerMeta, error) {
+	if g, ok := rt.RoundTripper.(PeerMetadataGetter); ok {
+		return g.GetPeerMetadata()
+	}
+
+	return nil, fmt.Errorf("can not get peer protocol map. Inner roundtripper does not implement GetPeerMetadata")
+}
+
+// RoundTrip implements http.RoundTripper.
+func (rt *namespacedRoundTripper) RoundTrip(r *http.Request) (*http.Response, error) {
+	if !strings.HasPrefix(r.URL.Path, rt.protocolPrefix) {
+		r.URL.Path = rt.protocolPrefix + r.URL.Path
+	}
+	if !strings.HasPrefix(r.URL.RawPath, rt.protocolPrefixRaw) {
+		r.URL.RawPath = rt.protocolPrefixRaw + r.URL.Path
+	}
+
+	return rt.RoundTripper.RoundTrip(r)
+}
+
+// NamespaceRoundTripper returns an http.RoundTripper that are scoped to the given protocol on the given server.
+func (h *Host) NamespaceRoundTripper(roundtripper http.RoundTripper, p protocol.ID, server peer.ID) (*namespacedRoundTripper, error) {
+	protos, err := h.getAndStorePeerMetadata(roundtripper, server)
+	if err != nil {
+		return &namespacedRoundTripper{}, err
+	}
+
+	v, ok := protos[p]
+	if !ok {
+		return &namespacedRoundTripper{}, fmt.Errorf("no protocol %s for server %s", p, server)
+	}
+
+	path := v.Path
+	if path[len(path)-1] == '/' {
+		// Trim the trailing slash, since it's common to make requests starting with a leading forward slash for the path
+		path = path[:len(path)-1]
+	}
+
+	u, err := url.Parse(path)
+	if err != nil {
+		return &namespacedRoundTripper{}, fmt.Errorf("invalid path %s for protocol %s for server %s", v.Path, p, server)
+	}
+
+	return &namespacedRoundTripper{
+		RoundTripper:      roundtripper,
+		protocolPrefix:    u.Path,
+		protocolPrefixRaw: u.RawPath,
+	}, nil
+}
+
+// NamespacedClient returns an http.Client that is scoped to the given protocol
+// on the given server. It creates a new RoundTripper for each call. If you are
+// creating many namespaced clients, consider creating a round tripper directly
+// and namespacing the roundripper yourself, then creating clients from the
+// namespace round tripper.
+func (h *Host) NamespacedClient(p protocol.ID, server peer.AddrInfo, opts ...RoundTripperOption) (http.Client, error) {
+	rt, err := h.NewRoundTripper(server, opts...)
+	if err != nil {
+		return http.Client{}, err
+	}
+
+	nrt, err := h.NamespaceRoundTripper(rt, p, server.ID)
+	if err != nil {
+		return http.Client{}, err
+	}
+
+	return http.Client{Transport: nrt}, nil
+}
+
+// NewRoundTripper returns an http.RoundTripper that can fulfill and HTTP
+// request to the given server. It may use an HTTP transport or a stream based
+// transport. It is valid to pass an empty server.ID.
+func (h *Host) NewRoundTripper(server peer.AddrInfo, opts ...RoundTripperOption) (http.RoundTripper, error) {
```

Comment:
I'm struggling with this abstraction. An `http.RoundTripper` is supposed to be able to issue requests to any HTTP host. This is more than just a theoretical concern: The `http.Client` has a `Transport` field that takes an `http.RoundTripper`, and would then be able to follow redirects. If we return a `RoundTripper` here that's scoped a single host, it won't be possible to redirect from one libp2p+HTTP host to another.

I don't have a good solution for this now, I'll have to think about the correct abstraction more.

Reply:
> it won't be possible to redirect from one libp2p+HTTP host to another.

I agree this is an interesting use case we should try to support.

The thing I like about this api is that a user can pass in many multiaddrs and the best transport is chosen for them. There isn't anywhere else we could do this besides when we create a round tripper. But, thinking more about this, this doesn't have to be a property of the core round tripper. We can have a round tripper on top of some core round tripper that can do this logic. It would be great if the core round tripper could, as you say, fulfill any HTTP request regardless of the transport.

Here's one idea how to do this:

1. Change this function to accept no parameters. It just returns an http.RoundTripper.
2. This roundtripper can fulfill HTTP requests to the following URIs:
  a. `http://` and `https://`, just like the http.DefaultTransport does.
  b. `multiaddr:<peer-multiaddr>`. Which would be a new URI scheme for multiaddrs. The value after `multiaddr:` is simply the string encoded version of the multiaddr. The multiaddr can include a peer's transport information, or could include only the peer id `/p2p/<peer-id>` and rely on discovery or out-of-band knowledge of a peer's addresses. Note this multiaddr can also be an http transport multiaddr.

This would work but would lead to us having to define the [httppath component](https://github.com/libp2p/specs/issues/554) since we'll need a way to reference the HTTP path we want to use in the multiaddr.

We would probably also want to register the `multiaddr` uri scheme in IANA (but could defer).

Here's a branch of this idea: https://github.com/libp2p/go-libp2p/compare/marco/http...marco/newroundtripper?expand=1. It includes an httppath implementation, but we should standardize that (not use the private area) if we go that route. It also has a test demonstrating the redirect to other host (using multiaddrs).

Reply:
For now let’s remove the NewRoundtripper and only expose the NamespacedRoundtripper and NamespacedClient. We can then add the NewRoundTripper later. 

The only thing we loose is redirects across hosts. I thinks that’s fine to omit for this initial version. And we have plan above on how to add it.

Reply:
I've renamed this to `NewConstrainedRoundTripper`, I think that will unblock us here. In a future version we can introduce the `NewRoundTripper` that accepts a multiaddr uri and can handle redirects across hosts.